### ISSUE TITLE
How to commit inisde plv8 procedure?

### ISSUE THREAD

**Issue by viditkapil** [plv8/plv8js#419]
In postgres 11 a new feature was intoduced where you can commit inside a procedure. I have tried this in plpgsql it works. But when i tried the same thing with plv8 it threw an error: **current transaction is aborted, commands ignored until end of transaction block.** 

Here is the sample code to test a plv8 procedure:

DROP PROCEDURE IF EXISTS test_extension_v(json) CASCADE;
CREATE OR REPLACE PROCEDURE test_extension_v(params json) AS $$
    plv8.elog(NOTICE, 'inside test_extension_v ', JSON.stringify(params));
    var response;
    for (var i = 0; i < 5; i++) {
        plv8.execute('select pg_sleep(5)');
        plv8.execute('insert into test (attributes) values ($1)', [params]);
        plv8.execute('commit');
        plv8.elog(NOTICE, 'commit: ', i);
    }
$$ LANGUAGE plv8;

psql=# create table test (id bigserial, attributes json);
psql=#call test_extension_v('{}');

It will throw error like this:
ERROR:  current transaction is aborted, commands ignored until end of transaction block
CONTEXT:  test_extension_v() LINE 8:         plv8.execute('commit');

But if you run the same code in plpgsql:

DROP PROCEDURE IF EXISTS test_extension_pl(json) CASCADE;
CREATE OR REPLACE PROCEDURE test_extension_pl(params json) language plpgsql AS $$
DECLARE
    response json;
begin
    for counter in 1..5 loop
        perform pg_sleep(5);
        insert into test (attributes) values (params);
        commit;
        raise notice 'commit: %', counter;
    end loop;
end; $$;

psql=#call test_extension_v('{}');

This will work as expected.

**Is it even possible to commit inside a plv8 procedure? if yes then how to use it and if not will this be supported in future.**

**Comment by JerrySievert** [comment]
I went through my notes from the pg10 to pg11 procedure changes, and here's where it stands:

plv8 supports `create procedure` and subtransactions, but does not currently directly support `commit`/`rollback`.  you can use the [`plv8.subtransaction`](https://plv8.github.io/#-code-plv8-subtransaction-code-) api if you need something similar.

there has been no call for this feature until now, so likely wouldn't be implemented unless there were more requests.  if it were implemented, it would use the `plv8.commit()` and `plv8.rollback()` pattern that pl/python uses.

I'll leave this issue open in case there are more requests for this feature.

**Comment by luss** [comment]
I add my +1 that this is a nice to have feature, but, I also agree that
this is non urgent for plv8.    I like and have used this new feature in
PLpgSQL, but, personally when this kinda of need presents itself I usually
try to run this kind of of "batch" logic in an app server rather than
inside of a database.  Commit/Rollback inside of stored procs is very
common for legacy apps written in PL/SQL for Oracle.    These things are
usually, in my experience, ported to PLpgSQL.

PLv8 is for writing kewl; new style code in a safe, sandboxed, & fast
environment.  :-)

**Comment by viditkapil** [comment]
> I went through my notes from the pg10 to pg11 procedure changes, and here's where it stands:
> 
> plv8 supports `create procedure` and subtransactions, but does not currently directly support `commit`/`rollback`. you can use the [`plv8.subtransaction`](https://plv8.github.io/#-code-plv8-subtransaction-code-) api if you need something similar.
> 
> there has been no call for this feature until now, so likely wouldn't be implemented unless there were more requests. if it were implemented, it would use the `plv8.commit()` and `plv8.rollback()` pattern that pl/python uses.
> 
> I'll leave this issue open in case there are more requests for this feature.

Subtransaction won't work in my case as i want to commit changes done so far, so that changes are visible to other clients too. 

Thanks for replying

**Comment by pkit** [comment]
The feature is new and implementation is pretty tricky, see the discussion here: https://github.com/YugaByte/yugabyte-db/issues/1957

**Comment by JerrySievert** [comment]
@viditkapil check out the branch `proc` of https://github.com/JerrySievert/plv8

note, that the usage pattern is:

```
CREATE PROCEDURE transaction_test1()
LANGUAGE plv8
AS $$
for (let i = 0; i < 10; i++) {
  plv8.execute('INSERT INTO test1 (a) VALUES ($1)', [i]);
  if (i % 2 == 0)
    plv8.commit();
  else
    plv8.rollback();
}
$$;
```

you cannot call `rollback` or `commit` inside of `plv8.execute()`.

**Comment by JerrySievert** [comment]
also note that this is a PoC - there are some changes I'd need to make before merging it into the core plv8 code.

**Comment by luss** [comment]
very kewl. +1

**Comment by JerrySievert** [comment]
I went ahead and finished up my commit, added tests, and merged it in to 3.0.

note that I don't plan on backporting this to the 2.3 branch at this time.

I'll update the documentation in the next day or so.

**Comment by luss** [comment]
That's great Jerry!!  It makes sense 2 me that this be a new feature of v3
and NOT backported.

**Comment by viditkapil** [comment]
> I went ahead and finished up my commit, added tests, and merged it in to 3.0.
> 
> note that I don't plan on backporting this to the 2.3 branch at this time.
> 
> I'll update the documentation in the next day or so.

Great, thank you for accepting my request and spending time on this feature. Will try and share my experience.

**Comment by viditkapil** [comment]
> > I went ahead and finished up my commit, added tests, and merged it in to 3.0.
> > note that I don't plan on backporting this to the 2.3 branch at this time.
> > I'll update the documentation in the next day or so.
> 
> Great, thank you for accepting my request and spending time on this feature. Will try and share my experience.

Perfect, I can use this in my project. Appreciate your quick response for helping me out with this feature.
